protocol/auditlog.go

```diff
@@ -0,0 +1,202 @@
+// Insert() returns an ErrAuditLog if the auditor attempts to create
+// a new history for a known directory, an ErrMalformedDirectoryMessage
+// if oldSTRs is malformed, a CheckBadSignature or CheckBadSTR if there
+// is an inconsistency in the history given in hist, and nil otherwise.
+// Insert() only creates the initial entry in the log for addr. Use Update()
+// to insert newly observed STRs for addr in subsequent epochs.
+// FIXME: pass Response message as param
+// masomel: will probably want to write a more generic function
+// for "catching up" on a history in case an auditor misses epochs
+func (l ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	hist map[uint64]*DirSTR) error {
+
+	// make sure we're getting an initial STR at the very least
+	if len(hist) < 1 && hist[0].Epoch != 0 {
+		return ErrMalformedDirectoryMessage
+	}
+
+	// compute the hash of the initial STR
+	dirInitHash := ComputeDirectoryIdentity(hist[0])
+
+	// error if we want to create a new entry for a directory
+	// we already know
+	if l.IsKnownDirectory(dirInitHash) {
+		return ErrAuditLog
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(addr, signKey, hist[0])
+
+	// add each STR into the history
+	// start at 1 since we've inserted the initial STR above
+	startEp := uint64(1)
+	endEp := uint64(len(hist))
+
+	// This loop automatically catches if hist is malformed
+	// (i.e. hist is missing an epoch between 0 and the latest given)
+	for ep := startEp; ep < endEp; ep++ {
+		str := hist[ep]
+		if str == nil {
+			return ErrMalformedDirectoryMessage
+		}
+
+		// verify the consistency of each new STR before inserting
+		// into the audit log
+		err := verifySTRConsistency(signKey, h.snapshots[ep-1], str)
+
+		if err != nil {
+			return err
+		}
+
+		h.snapshots[ep] = str
+	}
+
+	// Make sure to update the latestSTR
+	// in this particular call, the latestSTR has already been
+	// inserted into the snapshots map in the loop above
+	h.updateLatestSTR(hist[endEp-1])
+	l[dirInitHash] = h
+
+	return nil
+}
+
+// Update verifies the consistency of a newly observed STR newSTR for
+// the directory addr, and inserts the newSTR into addr's directory history
+// if the checks (i.e. STR signature and hash chain verifications) pass.
+// Update() returns nil if the checks pass, and the appropriate consistency
+// check error otherwise. Update() assumes that Insert() has been called for
+// addr prior to its first call and thereby expects that an entry for addr
+// exists in the audit log l.
+// FIXME: pass Response message as param
+func (l ConiksAuditLog) Update(dirInitHash string, newSTR *DirSTR) error {
+
+	// error if we want to update the entry for an addr we don't know
+	if !l.IsKnownDirectory(dirInitHash) {
+		return ErrAuditLog
+	}
+
+	h := l[dirInitHash]
+
+	if err := verifySTRConsistency(h.signKey, h.latestSTR, newSTR); err != nil {
+		return err
+	}
+
+	// update the latest STR
+	h.updateLatestSTR(newSTR)
```

Comment:
Because of what I'm doing for #170, I need to see if it makes sense to refactor these operations both on the client and auditor. I'll probably hold off on that for now.